cgp/individual.py

```diff
@@ -31,12 +32,17 @@ def __init__(self, fitness: Union[float, None]) -> None:
             Fitness of the individual.
         """
         self.fitness: Union[float, None] = fitness
-        self.idx: int
-        self.parent_idx: int
+        self.idx: Union[int, None] = None
+        self.parent_idx: Union[int, None] = None
 
     def clone(self):
         raise NotImplementedError()
 
+    def _copy_user_defined_attributes(self, other):
+        for attr in self.__dict__:
```

Comment:
If you copy everything that is in `self.__dict__`, then these attributes are not really "user-defined", are they?

Reply:
yeah, maybe there's a comment missing here. i iterate through `__dict__` but then exclude all members that would be present in a "clean" instance in the line below. so "user defined" means all members that are not defined at `__init__`.

Reply:
Yes, please add a comment. Is this the "proper" way to achieve this kind of behavior? It looks a bit "hacky" to me.

Reply:
hmm, i'm not sure what you mean by "proper", but it seems like a reasonable approach to me ;)

Reply:
is `self.__class__(None, None)` creating a new instance? If yes, you're creating a new instance in every function call? Isn't that a bit wasteful?

It seems that this feature could be better solved in another way with a member function that is executed by a callback that the user activates by some optional arguments in the high-level API?

Reply:
i agree, it's a bit wasteful, but clone is not called too often so i don't think this would be a big issue.

yes, you might be right. could you expand a bit on your suggestion? i will look more into multiprocessing since i'm not sure how easy it is to get variables back out of the parallel context. it is possible that this could only be done via return values.

Reply:
What I mean is that the `ea.MuPlusLambda._create_new_offspring_generation()` function could accept a callback that is called just after cloning: https://github.com/Happy-Algorithms-League/hal-cgp/blob/7932565d5112135a3c538d4ff6b948ba3020eb1b/cgp/ea/mu_plus_lambda.py#L136

This callback then sets attributes in the newly cloned individuals.
Alternatively, to be more restrictive: Just accept a set of attribute names as an arguments to the function which are then copied in the function from the parents to the cloned offspring.

Reply:
ok, made a suggestion, please have a look :)